Experimental features configuration and validation

slinkydeveloper opened the pull request:
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

Details: https://docs.google.com/document/d/1AoH0FyLeuHIg5snrlCKzELQv-NEA6bbjTU2Qv3zlW5k/edit?usp=sharing

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- :gift: Store to read the experimental features flags from a specific config map
- :gift: Validator for APIs when using experimental features
- :gift: Scaffolded the directory to add experimental features e2e/conformance tests

<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note

```

**Docs**

<!--
:book: If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

matzew commented:
How do I register "experimental features"?

slinkydeveloper commented:
> How do I register "experimental features"?

There is no need to "register" an experimental feature:

* By default, all features are disabled.
* If the CM contains `my-feature: "true"`, then the feature is enabled.

matzew commented:
> If the CM contains my-feature: "true", then the feature is enabled.

not fully getting how the `my-feature` is than really defined, as like more concrete:  timeout-enabeld?  

how would that apply to a timeout field?

slinkydeveloper commented:
@matzew the field still has to be in the api data structure, but it will be rejected by the webhook if the feature is not enabled. I'll explain it in the doc I'm writing :smile:

lionelvillard commented on config/400-config-experimental-features.yaml:
```diff
@@ -0,0 +1 @@
+core/configmaps/experimental-features.yaml
```
Can we call this just `features.yaml`, to be consistent with Knative Serving? A feature can be Alpha, ie experimental. A feature can also be stable and still be disabled if a vendor doesn't want it for X reason.

slinkydeveloper replied:
I added a comment in the doc about that

lionelvillard commented on config/core/configmaps/experimental-features.yaml:
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-experimental-features
```
same here

lionelvillard commented on pkg/apis/experimental/api_validation.go:
```diff
@@ -0,0 +1,34 @@
+package experimental
```
license

lionelvillard commented on pkg/apis/experimental/api_validation.go:
```diff
@@ -0,0 +1,34 @@
+
+			if (fieldVal.Kind() == reflect.Ptr || fieldVal.Kind() == reflect.Slice || fieldVal.Kind() == reflect.Map) && !fieldVal.IsNil() {
+				errs = errs.Also(&apis.FieldError{
+					Message: fmt.Sprintf("Disallowed field because the experimental feature '%s' is disabled", featureName),
```
remove experimental

slinkydeveloper commented:
I've added a directory (and modified the GH action) where we can place the experimental features e2e/conformance tests. These are not going to cause any troubles to existing e2e/conformance tests.

vaikas commented:
/lgtm
Thanks for driving this through. Will be great to keep on experimenting and improving and learning!
I'll wait on actual approval, since you said you'd do that tmw.

matzew reviewed (approved):
/lgtm
/approve
